libnetwork/libnetwork_linux_test.go

```diff
@@ -906,160 +843,133 @@ func TestResolvConf(t *testing.T) {
 		err error
 	)
 
-	t.Parallel()
+	teardown, err := pt.osctx.Set()
+	if err != nil {
+		return err
+	}
+	defer teardown(t)
 
-	pTest := flag.Lookup("test.parallel")
-	if pTest == nil {
-		t.Skip("Skipped because test.parallel flag not set;")
+	epName := fmt.Sprintf("pep%d", thrNumber)
+
+	if thrNumber == 1 {
+		ep, err = pt.net1.EndpointByName(epName)
+	} else {
+		ep, err = pt.net2.EndpointByName(epName)
 	}
-	numParallel, err := strconv.Atoi(pTest.Value.String())
+
 	if err != nil {
-		t.Fatal(err)
+		return errors.WithStack(err)
 	}
-	if numParallel < numThreads {
-		t.Skip("Skipped because t.parallel was less than ", numThreads)
+	if ep == nil {
+		return errors.New("got nil ep with no error")
 	}
 
-	runtime.LockOSThread()
-	if thrNumber == first {
-		createGlobalInstance(t)
-	} else {
-		<-start
-
-		thrdone := make(chan struct{})
-		done <- thrdone
-		defer close(thrdone)
+	cid := fmt.Sprintf("%drace", thrNumber)
+	pt.controller.WalkSandboxes(libnetwork.SandboxContainerWalker(&sb, cid))
+	if sb == nil {
+		return errors.Errorf("got nil sandbox for container: %s", cid)
+	}
 
-		if thrNumber == last {
-			defer close(done)
+	for i := 0; i < pt.iterCnt; i++ {
+		if err := ep.Join(sb); err != nil {
+			if _, ok := err.(types.ForbiddenError); !ok {
+				return errors.Wrapf(err, "thread %d", thrNumber)
+			}
 		}
-
-		err = netns.Set(testns)
-		if err != nil {
-			runtime.UnlockOSThread()
-			t.Fatal(err)
+		if err := ep.Leave(sb); err != nil {
+			if _, ok := err.(types.ForbiddenError); !ok {
+				return errors.Wrapf(err, "thread %d", thrNumber)
+			}
 		}
 	}
-	defer func() {
-		if err := netns.Set(origins); err != nil {
-			t.Fatalf("Error restoring the current thread's netns: %v", err)
-		} else {
-			runtime.UnlockOSThread()
-		}
-	}()
 
-	net1, err := controller.NetworkByName("testhost")
-	if err != nil {
-		t.Fatal(err)
-	}
-	if net1 == nil {
-		t.Fatal("Could not find testhost")
+	if err := errors.WithStack(sb.Delete()); err != nil {
+		return err
 	}
+	return errors.WithStack(ep.Delete(false))
+}
 
-	net2, err := controller.NetworkByName("network2")
-	if err != nil {
-		t.Fatal(err)
-	}
-	if net2 == nil {
-		t.Fatal("Could not find network2")
-	}
+func TestParallel(t *testing.T) {
+	const (
+		first      = 1
+		last       = 3
+		numThreads = last - first + 1
+		iterCnt    = 25
+	)
 
-	epName := fmt.Sprintf("pep%d", thrNumber)
+	osctx := testutils.SetupTestOSContextEx(t)
+	defer osctx.Cleanup(t)
+	controller := newController(t)
 
-	if thrNumber == first {
-		ep, err = net1.EndpointByName(epName)
-	} else {
-		ep, err = net2.EndpointByName(epName)
+	netOption := options.Generic{
+		netlabel.GenericData: options.Generic{
+			"BridgeName": "network",
+		},
 	}
 
+	net1 := makeTesthostNetwork(t, controller)
+	defer net1.Delete()
+	net2, err := createTestNetwork(controller, "bridge", "network2", netOption, nil, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
-	if ep == nil {
-		t.Fatal("Got nil ep with no error")
-	}
+	defer net2.Delete()
```

Comment:
Same applies for these (i.e., we need to use `defer`, and we cannot use `t.Cleanup()` (which could be the controller cleaning up all networks it created)?

Reply:
Yes: deleting network resources tends to not go very well when the execution context is in the wrong network namespace. Even if we ignore the arbitrary-goroutine problem, any registered `t.Cleanup()`s will be invoked after the test function returns, and therefore after `defer osctx.Cleanup(t)` is invoked.